h/static/scripts/build-thread.js

```diff
@@ -0,0 +1,295 @@
+'use strict';
+
+/** Default state for new threads, before applying filters etc. */
+var DEFAULT_THREAD_STATE = {
+  /** The Annotation which is displayed by this thread. */
+  annotation: undefined,
+  /** The parent Thread */
+  parent: undefined,
+  /** True if replies to this annotation are hidden. */
+  collapsed: false,
+  /** True if this annotation matches the current filters. */
+  visible: true,
+  /** Replies to this annotation. */
+  children: [],
+  /**
+    * The total number of children of this annotation,
+    * including any which have been hidden by filters.
+    */
+  totalChildren: 0,
+};
+
+/**
+ * Returns a persistent identifier for an Annotation.
+ * If the Annotation has been created on the server, it will have
+ * an ID assigned, otherwise we fall back to the local-only '$$tag'
+ * property.
+ */
+function id(annotation) {
+  return annotation.id || annotation.$$tag;
+}
+
+/**
+ * Creates a thread of annotations from a list of annotations.
```

Comment:
This comment should certainly explain where the core of the algorithm comes from 😉